test/test_external.py

```diff
@@ -331,5 +339,61 @@ def test_can_load_tabular_model_with_different_widget_data(hypothetical_readme):
         check_dataset(io.config, hypothetical_readme)
 
 
+@pytest.mark.parametrize(
+    "config, answer",
+    [
+        ({"version": "3.3", "enable_queue": True}, True),
+        ({"version": "3.3", "enable_queue": False}, False),
+        ({"version": "3.2", "enable_queue": False}, False),
+        ({"version": "3.2", "enable_queue": True}, True),
+        ({"version": "3.1.3", "enable_queue": True}, False),
+        ({"version": "3.1.3", "enable_queue": False}, False),
+    ],
+)
+def test_use_websocket_after_315(config, answer):
+    assert use_websocket(config) == answer
+
+
+class AsyncMock(MagicMock):
+    async def __call__(self, *args, **kwargs):
+        return super(AsyncMock, self).__call__(*args, **kwargs)
+
+
+@pytest.mark.asyncio
+async def test_get_pred_from_ws():
+    mock_ws = AsyncMock(name="ws")
+    messages = [
+        json.dumps({"msg": "estimation"}),
+        json.dumps({"msg": "send_data"}),
+        json.dumps({"msg": "process_generating"}),
+        json.dumps({"msg": "process_completed", "output": {"data": ["result!"]}}),
+    ]
+    mock_ws.recv.side_effect = messages
+    data = json.dumps({"data": ["foo"], "fn_index": "foo"})
+    output = await get_pred_from_ws(mock_ws, data)
+    assert output == {"data": ["result!"]}
+    mock_ws.send.assert_called_once_with(data)
+
+
+@pytest.mark.asyncio
+async def test_get_pred_from_ws_raises_if_queue_full():
+    mock_ws = AsyncMock(name="ws")
+    messages = [json.dumps({"msg": "queue_full"})]
+    mock_ws.recv.side_effect = messages
+    data = json.dumps({"data": ["foo"], "fn_index": "foo"})
+    with pytest.raises(gradio.Error, match="Queue is full!"):
+        await get_pred_from_ws(mock_ws, data)
+
+
+@pytest.mark.skipif(sys.version_info < (3, 8), reason="Async mocks don't work for 3.7")
```

Comment:
I'm confused, as there's as `AsyncMock` in the previous tests which are not skipped?

Reply:
The problem is that `websockets.connect` returns an async context manager and mock doesn't really work with async context managers in 3.7. I can make the `reason` a bit more precise.

Reply:
Got it thanks for clarifying! Good to know